[libc++] Implement `bind_back`

JMazurkiewicz opened the pull request:
Implement `std::bind_back` function from P2387R3 "Pipe support for user-defined range adaptors".

frederick-vs-ja commented on libcxx/test/std/utilities/function.objects/func.bind.partial/bind_back.pass.cpp:
```diff
@@ -0,0 +1,356 @@
+  {
+    static_assert(!std::is_constructible_v<NotCopyMove, NotCopyMove&>);
+    static_assert(!std::is_move_constructible_v<NotCopyMove>);
+    static_assert(!back_bindable<NotCopyMove>);
+    static_assert(!back_bindable<NotCopyMove&>);
```
Perhaps we should use `LIBCPP_STATIC_ASSERT` for libc++-specific SFINAE-friendliness, since the standard uses _Mandates_ for unsuitable parameter types ([[func.bind.partial]/2](https://eel.is/c++draft/func.bind.partial#2)).

Ditto for other occurences of `back_bindable` and pre-existing test cases for `bind_front`.

```suggestion
    LIBCPP_STATIC_ASSERT(!back_bindable<NotCopyMove>);
    LIBCPP_STATIC_ASSERT(!back_bindable<NotCopyMove&>);
```

JMazurkiewicz committed:
[libc++] Implement `bind_back`

Implement `std::bind_back` function from P2387R3 "Pipe support for user-defined range adaptors".

JMazurkiewicz committed:
Clang modules: Fix build

ldionne commented on libcxx/test/std/utilities/function.objects/func.bind.partial/bind_back.verify.cpp:
```diff
@@ -0,0 +1,80 @@
+    auto p = std::bind_back(pass, 1);
+    static_assert(p() == 1); // expected-error {{static assertion expression is not an integral constant expression}}
+
+    auto d = std::bind_back(do_nothing, 2); // expected-error {{no matching function for call to 'bind_back'}}
```
What is this test checking? You can't pass an overloaded function as an argument. Am I missing something, or this test isn't testing anything from the library?

JMazurkiewicz replied:
This test checks if there exists some (incorrect) overload of `std::bind_back`. I'm going to add comment with clarification.

> You can't pass an overloaded function as an argument.

You can do something like [this](https://godbolt.org/z/xzcEcGcr5):
```c++
namespace std {
    template<class F, class... Args>
    auto bind_back(F&&, Args&&...);

    void bind_back(int (*)(int));
}

template<class T>
T do_nothing(T t) {
    return t;
}

void use() {
    std::bind_back(do_nothing);
}
```

BTW, I've copied this test from `bind_front.verify.cpp`:
https://github.com/llvm/llvm-project/blob/1c3b15e9f5bc671e40bcf5d3475f5425466754ce/libcxx/test/std/utilities/function.objects/func.bind_front/bind_front.verify.cpp#L42

ldionne commented on libcxx/test/std/utilities/function.objects/func.bind.partial/bind_back.verify.cpp:
```diff
@@ -0,0 +1,80 @@
+
+void test() {
+  { // Various failures
+    auto p = std::bind_back(pass, 1);
```
If it's fundamental that `pass` is an actual function (and not a function object), I would mention it in a comment and I would define `pass` in this test file so we can see its definition immediately.

JMazurkiewicz replied:
Done, commits: https://github.com/llvm/llvm-project/pull/81055/commits/66632baa67282869fd0de13dd0144ed8a8df7be1, https://github.com/llvm/llvm-project/pull/81055/commits/5d68323a202f4fee6d4ce5beb132e8759732741c

ldionne commented on libcxx/test/std/utilities/function.objects/func.bind.partial/bind_back.pass.cpp:
```diff
@@ -0,0 +1,362 @@
+#include "callable_types.h"
+#include "types.h"
+
+constexpr void test_basic_bindings() {
```
Do we have any test that we can return e.g. a reference from the function and it won't get decayed? I couldn't find any, but maybe I missed them.

If we don't have any, we should also add a similar test for `bind_front` while we're at it, it's pretty easy to test.

JMazurkiewicz replied:
Done:
* `bind_back`: https://github.com/llvm/llvm-project/pull/81055/commits/730b4e72184ce9a1914300b86e8748ab960c5122
* `bind_front`: https://github.com/llvm/llvm-project/pull/81055/commits/88013f8f222dfdd111b950d5c4496616e03f3aa4

ldionne reviewed (changes requested):
Thanks for the patch! I have some comments but this is looking pretty good.

JMazurkiewicz committed:
Move `pass` to `bind_back.verify.cpp` and add better comment

Addresses https://github.com/llvm/llvm-project/pull/81055#discussion_r1521919109

JMazurkiewicz committed:
Test returning reference from the function (`bind_back`)

Addresses https://github.com/llvm/llvm-project/pull/81055#discussion_r1521924160

JMazurkiewicz committed:
Explain passing template function as an argument

Addresses (maybe) https://github.com/llvm/llvm-project/pull/81055#discussion_r1521918117

JMazurkiewicz committed:
Test returning reference from the function (`bind_front`)

Addresses https://github.com/llvm/llvm-project/pull/81055#discussion_r1521924160

JMazurkiewicz committed:
Merge branch 'main' into libcxx/functional/bind_back

JMazurkiewicz committed:
Better comment

JMazurkiewicz committed:
Fix CI (Clang modules)

JMazurkiewicz committed:
Fix CI (C++20 failure due to copy-paste)

JMazurkiewicz committed:
Merge branch 'main' into libcxx/functional/bind_back

JMazurkiewicz committed:
Merge branch 'main' into libcxx/functional/bind_back

ldionne commented:
Thanks for the patch! I'll merge now since it's green.
